src/D2L.CodeStyle.Analyzers/ApiUsage/Serialization/ReflectionSerializerAnalyzer.cs

```diff
@@ -0,0 +1,456 @@
+					& c.Modifiers.IndexOf( SyntaxKind.StaticKeyword ) < 0
+				) )
+				.ToImmutableArray();
+
+			return constructors;
+		}
+
+		private static bool IsPublicInstanceConstructor( IMethodSymbol method ) {
+
+			if( method.MethodKind != MethodKind.Constructor ) {
+				return false;
+			}
+
+			if( method.DeclaredAccessibility != Accessibility.Public ) {
+				return false;
+			}
+
+			return true;
+		}
+
+		private static void AnalyzePublicInstanceConstructors(
+				SyntaxNodeAnalysisContext context,
+				ReflectionSerializerModel model,
+				TypeDeclarationSyntax type,
+				TypeDiagnostics diagnostics
+			) {
+
+			INamedTypeSymbol typeSymbol = context.SemanticModel.GetDeclaredSymbol( type );
+
+			ImmutableArray<IMethodSymbol> constructors = typeSymbol
+				.GetMembers()
+				.OfType<IMethodSymbol>()
+				.Where( IsPublicInstanceConstructor )
+				.ToImmutableArray();
+
+			switch( constructors.Length ) {
+
+				case 0:
+					ReportDiagnostic(
+							context,
+							diagnostics.NoPublicConstructor,
+							type
+						);
+					break;
+
+				case 1:
+					if( diagnostics.ConstructorParameterCannotBeDeserialized != null ) {
+						AnalyzeConstructorParameters(
+								context,
+								model,
+								type,
+								constructors[ 0 ].Parameters,
+								diagnostics.ConstructorParameterCannotBeDeserialized
+							);
+					}
+					break;
+
+				default:
+					ReportDiagnostic(
+							context,
+							diagnostics.MultiplePublicConstructors,
+							type
+						);
+					break;
+			}
+		}
+
+		/// <summary>
+		/// Analyzes the constructor parameters using the syntax in the event of a partial type,
+		/// where the [ReflectionSerializer] attribute and constructor are defined in the same file.
+		/// </summary>
+		private static void AnalyzeConstructorParameters(
+				SyntaxNodeAnalysisContext context,
+				ReflectionSerializerModel model,
+				TypeDeclarationSyntax type,
+				ParameterListSyntax constructorParameters,
+				DiagnosticDescriptor parameterCannotBeDeserializedDiagnostic
+			) {
+
+			INamedTypeSymbol typeSymbol = context.SemanticModel.GetDeclaredSymbol( type );
+			ImmutableHashSet<string> serializedPropertyNames = model.GetPublicReadablePropertyNames( typeSymbol );
+
+			foreach( ParameterSyntax parameter in constructorParameters.Parameters ) {
+
+				string parameterName = parameter.Identifier.ValueText;
+				if( !serializedPropertyNames.Contains( parameterName ) ) {
+
+					ReportConstructorParameterCannotBeDeserialized(
+							context,
+							descriptor: parameterCannotBeDeserializedDiagnostic,
+							identifier: parameter.Identifier,
+							parameterName: parameter.Identifier.ValueText
+						);
+				}
+			}
+		}
+
+		/// <summary>
+		/// Analyzes the constructor parameters using the symbols in the event of a partial type,
+		/// where the [ReflectionSerializer] attribute and constructor are not in the same file.
+		/// </summary>
+		private static void AnalyzeConstructorParameters(

```

Comment:
Haven't read anything in depth, but would it be beneficial to do everything at the symbolic level, using `RegisterSymbolAction` and `SymbolKind.NamedType`, and doing the initial check for `[ReflectionSerializer]` in `symbol.GetAttributes()`?

Reply:
Maybe?  That at least can point me at the different partial locations:

https://docs.microsoft.com/en-us/dotnet/api/microsoft.codeanalysis.isymbol.locations?view=roslyn-dotnet-4.0.1#Microsoft_CodeAnalysis_ISymbol_Locations